File: GNGradientView/UIView+Gradient.m
//
//  UIView+Gradient.m
//  16
//
//  Created by Jakub Knejzlik on 21/10/14.
//  Copyright (c) 2014 Jakub Knejzlik. All rights reserved.
//

#import "UIView+Gradient.h"

@implementation UIView (GNGradient)

-(CAGradientLayer *)gradientMaskLayer{
    if (![self.layer.mask isKindOfClass:[CAGradientLayer class]]) {
        self.layer.mask = [CAGradientLayer layer];
    }
    return (CAGradientLayer *)self.layer.mask;
}
-(void)updateGradientMaskFrame{
    self.layer.mask.frame = self.layer.bounds;
}


-(void)setGradientMaskAlphaChannels:(NSArray *)alphaChannels{
    NSMutableArray *colors = [NSMutableArray array];
    for (NSNumber *alpha in alphaChannels) {
        [colors addObject:(id)[[[UIColor whiteColor] colorWithAlphaComponent:[alpha floatValue]] CGColor]];
    }
    self.gradientMaskLayer.colors = colors;
}

-(void)setGradientMaskDirection:(GNGradientViewDirection)direction{
    switch (direction) {
        case GNGradientViewDirectionBottom:
            self.gradientMaskLayer.startPoint = CGPointMake(0.5, 0);
            self.gradientMaskLayer.endPoint = CGPointMake(0.5, 1);
            break;
        case GNGradientViewDirectionTop:
            self.gradientMaskLayer.startPoint = CGPointMake(0.5, 1);
            self.gradientMaskLayer.endPoint = CGPointMake(0.5, 0);
            break;
        case GNGradientViewDirectionLeft:
            self.gradientMaskLayer.startPoint = CGPointMake(1, 0.5);
            self.gradientMaskLayer.endPoint = CGPointMake(0, 0.5);
            break;
        case GNGradientViewDirectionRight:
            self.gradientMaskLayer.startPoint = CGPointMake(0, 0.5);
            self.gradientMaskLayer.endPoint = CGPointMake(1, 0.5);
            break;
    }
}

@end

File: GNGradientView/Classes/UIView+Gradient.swift
//
//  UIView+Gradient.swift
//  Pods
//
//  Created by Jakub Knejzlik on 19/09/16.
//
//

import UIKit

public enum GradientDirection {
    case vertical
    case horizontal
}

extension UIView {
    
    var gradientMaskLayer: CAGradientLayer {
        if let mask = self.layer.mask as? CAGradientLayer {
            return mask
        }
        let mask = CAGradientLayer()
        self.layer.mask = mask
        return mask
    }
    
    public func gradientView_setMaskDirection(value: GradientDirection) {
        switch value {
        case .vertical:
            gradientMaskLayer.startPoint = CGPoint(x: 0.5, y: 0)
            gradientMaskLayer.endPoint = CGPoint(x: 0.5, y: 1)
            break
        case .horizontal:
            gradientMaskLayer.startPoint = CGPoint(x: 0, y: 0.5)
            gradientMaskLayer.endPoint = CGPoint(x: 1, y: 0.5)
            break
        }
    }
    
    public func gradientView_setMaskAlphaChannels(_ channels: [CGFloat]) {
        gradientMaskLayer.colors = channels.map { UIColor(white: 1, alpha: $0).cgColor }
    }
    
    public func gradientView_updateMaskFrame() {
        self.layer.mask?.frame = self.layer.bounds
    }

}
